DNN Platform/Library/Entities/EventManager.cs

```diff
@@ -102,6 +102,10 @@ public EventManager(IEventLogger eventLogger)
                 this.RoleDeleted += handler.Value.RoleDeleted;
                 this.RoleJoined += handler.Value.RoleJoined;
                 this.RoleLeft += handler.Value.RoleLeft;
+                if (handler.Value is IRoleUpdateEventHandlers updateHandler)

```

Comment:
I would think you'd want to be able to implement `IRoleUpdateEventHandlers` without also implementing `IRoleEventHandlers`. If you only want to handle the `RoleUpdated` event, it would be surprising to me as a developer that I need to implement the other interface, too.

Reply:
To find the handler implementations, DNN uses `MefComposition` so an implementation would look something like this (from what I could figure out):
```cs
[Export(typeof(IRoleEventHandlers))]
[ExportMetadata("MessageType", "RoleEventHandlers")]
public class RoleEvents : IRoleEventHandlers {
    // ... implementation
}
```

If we go with `interface IRoleUpdateEventHandlers  : IRoleEventHandlers {...}` and then only implement that, I think that the final implementation would be something like this:

```cs
[Export(typeof(IRoleEventHandlers))]
[ExportMetadata("MessageType", "RoleEventHandlers")]
public class RoleEvents : IRoleUpdateEventHandlers {
    // ... implementation
}
```

Which, in my opinion, would be very hard to figure out that you need to implement `IRoleUpdateEventHandlers` but you still need to specify the export as the old interface `[Export(typeof(IRoleEventHandlers))]`, otherwise I would guess that the `EventHandlersContainer<IRoleEventHandlers>` [here](https://github.com/dnnsoftware/Dnn.Platform/blob/develop/DNN%20Platform/Library/Entities/EventManager.cs#L99) would not get it.

Reply:
I was thinking that the event manager would start using `EventHandlers<IRoleUpdateEventHandlers>` so that an implementation would be 

```
[Export(typeof(IRoleUpdateEventHandlers))]
[ExportMetadata("MessageType", "IRoleUpdateEventHandlers")]
public class RoleEvents : IRoleUpdateEventHandlers {
    // ... implementation
}
```

But I guess that makes it more difficult for the common scenario where you do want both.

We could support both scenarios (where `IRoleUpdateEventHandlers` is registered as `IRoleUpdateEventHandlers` or `IRoleUpdateEventHandlers` is registered as `IRoleEventHandlers`). I feel like that would be a more friendly dev experience. Open to discussion if others don't think that's worth it or makes it more complicated.

Reply:
So what you are saying is that we should do both things: a foreach on `EventHandlersContainer<IRoleUpdateEventHandlers>` and a cast check to `IRoleUpdateEventHandlers` in the existing `EventHandlersContainer<IRoleEventHandlers>` foreach, so that we support both scenarios.

I have no problem with that, but I am not sure what should happen next. Will you make a decision in the next approvers meeting?

Same for the comment below with the smaller/versioned interfaces.

_As a side note/question for the versioned interfaces or base classes: it will help with adding new things to the interfaces without breaking changes, but what will happen when you want to remove obsolete API from the interface and stop using/supporting that API throughout DNN's code? Will that just go through the normal two versions obsolete then removal process as a breaking change? Will the versioned interfaces be supported only for a specific number of major versions?_

Reply:
Yeah, I think we'll probably need to have a discussion at least on the interface question next Tuesday. We'll let you know what we decide. Thanks!

Reply:
@bdukes Have you had the chance to discuss about this?

Reply:
We did discuss this on Tuesday. This PR includes a breaking change on the `IEventManager` interface, and so will have to be held until at least version 10.0. We're planning to have a larger discussion in about a month around making these sorts of changes easier in the future.